notification-server/src/bin/api.rs

```diff
@@ -31,20 +40,75 @@ struct Args {
     log_level:      log::LevelFilter,
 }
 
-#[derive(Clone)]
+#[derive(Debug)]
 struct AppState {
-    db_connection: Config,
+    db_connection: DatabaseConnection,
+}
+
+const MAX_RESOURCES_LENGTH: usize = 1000;
+
+lazy_static! {
+    static ref MAX_PREFERENCES_LENGTH: usize = all::<Preference>().collect::<Vec<_>>().len();
 }
 
+#[tracing::instrument]
 async fn upsert_account_device(
     Path(device): Path<String>,
     State(state): State<Arc<AppState>>,
-    Json(account): Json<Vec<String>>,
-) -> Result<impl axum::response::IntoResponse, axum::response::Response> {
-    info!("Subscribing accounts {:?} to device {}", account, device);
-    let _ = &state.db_connection;
-    // TODO write to the database
-    Ok(StatusCode::OK)
+    Json(subscription): Json<DeviceSubscription>,
+) -> Result<impl IntoResponse, Response> {
+    info!(
+        "Subscribing accounts {:?} to device {}",
+        subscription, device
+    );
+    if subscription.preferences.len() > *MAX_PREFERENCES_LENGTH {
+        return Err((
+            StatusCode::BAD_REQUEST,
+            format!("Preferences exceeded length of {}", *MAX_PREFERENCES_LENGTH),
+        )
+            .into_response())?;
+    }
+    let unique_preferences: HashSet<_> = subscription.preferences.iter().collect();
+    if unique_preferences.len() != subscription.preferences.len() {
+        return Err((StatusCode::BAD_REQUEST, "Duplicate preferences found").into_response());
+    }
+
+    if subscription.accounts.len() > MAX_RESOURCES_LENGTH {
+        return Err((
+            StatusCode::BAD_REQUEST,
+            format!("Accounts exceed maximum length of {}", MAX_RESOURCES_LENGTH),
+        )
+            .into_response())?;
+    }
+    let decoded_accounts: Result<Vec<Vec<u8>>, Response> = subscription
+        .accounts
+        .iter()
+        .map(|account| {
+            let account_address: Result<AccountAddress, AccountAddressParseError> =
+                AccountAddress::from_str(account);
+            account_address
+                .map_err(|e| {
+                    error!("Failed to parse account address: {}", e);
+                    (StatusCode::BAD_REQUEST, "Failed to parse account address").into_response()
+                })
+                .map(|value| value.0.to_vec())
+        })
+        .collect();
+    let decoded_accounts = decoded_accounts?;
```

Comment:
Your current conversion is: `AccountAddress -> Bytes -> Vec<u8> -> Bytes -> Database`.

It can be simplified by `AccountAddress -> Bytes -> Database`

```suggestion
    let decoded_accounts: Result<Vec<AccountAddress>, Response> = subscription
        .accounts
        .iter()
        .map(|account| {
            AccountAddress::from_str(account).map_err(|e| {
                error!("Failed to parse account address: {}", e);
                (StatusCode::BAD_REQUEST, "Failed to parse account address").into_response()
            })
        })
        .collect();
    let decoded_accounts = decoded_accounts?;
```

```
    pub async fn upsert_subscription(
        &self,
        address: Vec<AccountAddress>,
        preferences: Vec<Preference>,
        device_id: &str,
    ) -> anyhow::Result<()> {
        let mut client = self
            .pool
            .get()
            .await
            .map_err(|e| anyhow!("Failed to get client: {}", e))?;
        let preferences_mask = preferences_to_bitmask(&preferences);
        let transaction = client.transaction().await?;
        for account in address {
            let params: &[&(dyn tokio_postgres::types::ToSql + Sync)] =
                &[&(account.0.as_ref()), &device_id, &preferences_mask];
            if let Err(e) = transaction.execute(&self.upsert_device, params).await {
                let _ = transaction.rollback().await;
                return Err(e.into());
            }
        }
        transaction.commit().await.map_err(Into::into)
    }
```

Reply:
True.